File: Leetcode/Easy Problems/1496. Path Crossing/dec-24-2023.js
/**
  December 24, 2023
  - Runtime Achieved: 50 ms
  - Memory Achieved: 42.1 MB
/**
 * @param {string} path
 * @return {boolean}
 */
var isPathCrossing = function(path) {
    let x=0, y=0;
    let visit = new Set();
    visit.add("0,0");
    for(let i of path) { 
        if(i.toUpperCase()==='N') y++;
        else if(i.toUpperCase()==='E') x++;
        else if(i.toUpperCase()==='W') x--;
        else if(i.toUpperCase()==='S') y--;

        let current = `${x},${y}`;
        if (visit.has(current)) {
            return true;
        } 
        else {
            visit.add(current);
        }
    }
    return false;
};

File: Leetcode/Easy Problems/1496. Path Crossing/dec-24-2023.cpp
/*
  December 24, 2023
  - Runtime Achieved: 4 ms 
  - Memory Achieved: 7.2 MB
*/
class Solution {
public:
    bool isPathCrossing(string path) {
        int x = 0, y = 0;
        unordered_set<string> visit;
        visit.insert("0,0");

        for(char direction : path) {
            if(toupper(direction) == 'N') y++;
            else if(toupper(direction) == 'E') x++;
            else if(toupper(direction) == 'W') x--;
            else if(toupper(direction) == 'S') y--;

            string current = to_string(x) + "," + to_string(y);
            if(visit.find(current) != visit.end()) {
                return true;
            } 
            else {
                visit.insert(current);
            }
        }
        return false;
    }
};
